src/github.ts

```diff
@@ -1940,9 +1964,132 @@ export class GitHub {
     const comparison = await this.octokit.repos.compareCommitsWithBasehead({
       owner: this.repository.owner,
       repo: this.repository.repo,
-      basehead: `${branchAName}...${branchBName}`,
+      basehead: `${branchBName}...${branchAName}`,
     });
-    return comparison.data.total_commits === 0;
+
+    this.logger.debug({status: comparison.data.status});
+
+    if (
+      comparison.data.status === 'identical' ||
+      comparison.data.status === 'ahead' ||
+      comparison.data.total_commits === 0
+    ) {
+      return true;
+    }
+
+    if (comparison.data.status === 'diverged') {
+      // For each branch fetch commits since branches diverged
+      const exclusiveCommitsA = new Set(
+        comparison.data.commits.map(commit => commit.sha)
+      );
+      const exclusiveCommitsB = new Set(
+        (
+          await this.octokit.repos.compareCommitsWithBasehead({
+            owner: this.repository.owner,
+            repo: this.repository.repo,
+            basehead: `${comparison.data.merge_base_commit.sha}...${branchBName}`,
+          })
+        ).data.commits.map(commit => commit.sha)
+      );
+
+      type CommitData = Awaited<
+        ReturnType<typeof this.octokit.repos.getCommit>
+      >['data'];
+      type CommitFiles = NonNullable<CommitData['files']>;
+
+      const getChangedFiles = (files: CommitFiles | undefined) =>
+        files?.map(file => ({
+          sha: file.sha,
+          filename: file.filename,
+          status: file.status,
+          additions: file.additions,
+          deletions: file.deletions,
+          changes: file.changes,
+          patch: file.patch,
+        })) || [];
+
+      // cache branch B data to avoid unnecessary requests
+      const cacheOfBranchBCommitMetadataPerCommitSha = new Map<
+        string,
+        CommitData['commit']
+      >();
+      const cacheOfBranchBFilesPerCommitSha = new Map<
+        string,
+        Map<string, ReturnType<typeof getChangedFiles>[0]>
+      >();
+
+      // Check if branch B seems to contain all commits exclusives to A
+      for (const shaA of exclusiveCommitsA) {
+        const commitA = await this.octokit.repos.getCommit({
+          owner: this.repository.owner,
+          repo: this.repository.repo,
+          ref: shaA,
+        });
+        const metaA = commitA.data.commit;
+        const filesA = getChangedFiles(commitA.data.files);
+
+        let foundInB = false;
+        for (const shaB of exclusiveCommitsB) {
+          let metaB = cacheOfBranchBCommitMetadataPerCommitSha.get(shaB);
```

Comment:
No need to make this a cache, you can just fill it out before hand

Reply:
Would it be easier to understand, if you read A and B into sets (with only the fields you need for comparison) and then do a set diff? I think that's what's happening here, but it's hard to grok

Reply:
> Would it be easier to understand, if you read A and B into sets (with only the fields you need for comparison) and then do a set diff? I think that's what's happening here, but it's hard to grok

That would be technically correct but may result in more requests, which is what I tried to minimize here. What you describe would always be `requestCount == len(commitsFromA) + len(commitsFromB)`, where the current implementation has `requestCount <= len(commitsFromA) + len(commitsFromB)` because we stop fetching commits from branch A as soon as we found one that isn't in B.

Reply:
> No need to make this a cache, you can just fill it out before hand

That would work too and be functionally equivalent but with a small technical difference: we would always fetch all exclusive commits from B, even when that's unnecessary. If fetching the first `N` commits from B is enough to find out that all exclusive commits from A are present, no need to send requests for the remaining commits.